src/main/java/de/tum/in/www1/artemis/service/messaging/services/LectureServiceProducer.java

```diff
@@ -0,0 +1,123 @@
+package de.tum.in.www1.artemis.service.messaging.services;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import javax.jms.JMSException;
+import javax.jms.Message;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jms.annotation.EnableJms;
+import org.springframework.jms.core.JmsTemplate;
+import org.springframework.stereotype.Component;
+import org.springframework.util.CollectionUtils;
+
+import de.tum.in.www1.artemis.config.MessageBrokerConstants;
+import de.tum.in.www1.artemis.domain.Course;
+import de.tum.in.www1.artemis.domain.Lecture;
+import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.service.dto.UserLectureDTO;
+import de.tum.in.www1.artemis.web.rest.errors.InternalServerErrorException;
+
+/**
+ * Message broker producer to send messages related to the exercise service
+ */
+@Component
+@EnableJms
+public class LectureServiceProducer {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LectureServiceProducer.class);
+
+    @Autowired
+    private final JmsTemplate jmsTemplate;
+
+    public LectureServiceProducer(JmsTemplate jmsTemplate) {
+        this.jmsTemplate = jmsTemplate;
+    }
+
+    /**
+     * Send a message to the Lecture microservice to filter the active attachments for a given user and wait for response.
+     *
+     * @param lectures the lectures
+     * @param user     the user
+     * @return the response from the Lecture microservice including set of lectures
+     */
+    public Set<Lecture> filterActiveAttachments(Set<Lecture> lectures, User user) {
+        if (CollectionUtils.isEmpty(lectures) || user == null) {
+            return new HashSet<>();
+        }
+        UserLectureDTO userLectureDTO = new UserLectureDTO(lectures, user);
+        LOGGER.info("Send message in queue {} with body {}", MessageBrokerConstants.LECTURE_QUEUE_FILTER_ACTIVE_ATTACHMENTS, userLectureDTO);
+        String correlationId = Integer.toString(userLectureDTO.hashCode());
+        jmsTemplate.convertAndSend(MessageBrokerConstants.LECTURE_QUEUE_FILTER_ACTIVE_ATTACHMENTS, userLectureDTO, message -> {
```

Comment:
The `message -> { … }` lambda is repeated quite a few times. Would it be possible to create a new class `ArtemisJmsTemplate extends JmsTemplate` that implements a few additional helper methods like
```java
public class ArtemisJmsTemplate extends JmsTemplate {
    public <T> void sendWithCorrelationId(String queue, String correlationId, T value) {
        convertAndSend(queue, value, msg -> {
            msg.setJMSCorrelationID(correlationId);
            return msg;
        });
    }

    public <T> T receiveWithCorrelationId(String queue, String correlationId, Class<T> bodyType) {
        try {
            Message message = receiveSelected(queue, "JMSCorrelationId='" + correlationId + "'");
            if (message != null) {
                return message.getBody(bodyType);
            } else {
                return null;
            }
        } catch (JMSException e) {
            throw new InternalServerErrorException("…");
        }
    }
}
```
to reduce duplication when sending and receiving the messages?

Additionally, turning the constants in `MessageBrokerConstants` into an enum and implementing a method there to retrieve the queue name as String could make the passed arguments to a potential `ArtemisJmsTemplate` more type-safe.


If it is not possible to extend the class, a utility class that provides something like
```java
public static MessagePostProcessor withCorrelationId(String correlationId) {
    return message -> {
        message.setJMSCorrelationID(correlationID);
        return message;
    };
}
```
would maybe reduce duplication. Then you could write `jmsTemplate.convertAndSend(MessageBrokerConstants.LECTURE_QUEUE_FILTER_ACTIVE_ATTACHMENTS, UtilityClass.withCorrelationId(correlationId));`

Reply:
I like those suggestions, I will implement them

Reply:
I tried to extend the JmsTemplate class, but unfortunately it didn't work. There were issues with the configuration of the connection to the broker. For this reason, I added a util class (JmsMessageUtil). 

Changing MessageBrokerConstants into an enum also didn't work out well, because we need constants in the consumer listeners. I tried to do it with constant defined as a subclass of the enum, but I think it was ugly, and I reverted the change :smile: